x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -144,229 +139,320 @@ public void authorize(Authentication authentication, String action, TransportReq
+                        // information such as the index and the incoming address of the request
+                        auditTrail.accessGranted(requestId, authentication, action, unwrappedRequest, authzInfo);
+                        listener.onResponse(null);
+                    }
+                } else {
+                    assert false :
+                        "only scroll related requests are known indices api that don't support retrieving the indices they relate to";
+                    listener.onFailure(denial(requestId, authentication, action, unwrappedRequest, authzInfo));
+                }
+            } else if (unwrappedRequest instanceof IndicesRequest &&
+                IndicesAndAliasesResolver.allowsRemoteIndices((IndicesRequest) unwrappedRequest)) {
+                // remote indices are allowed
+                final MetaData metaData = clusterService.state().metaData();
+                final AuthorizedIndices authorizedIndices = new AuthorizedIndices(
+                    () -> authzEngine.loadAuthorizedIndices(authentication, unwrappedRequest, action, authzInfo,
+                        metaData.getAliasAndIndexLookup()));
+                resolveIndexNames(unwrappedRequest, metaData, authorizedIndices, ActionListener.wrap(resolvedIndices -> {
+                    assert !resolvedIndices.isEmpty()
+                        : "every indices request needs to have its indices set thus the resolved indices must not be empty";
+                    //all wildcard expressions have been resolved and only the security plugin could have set '-*' here.
+                    //'-*' matches no indices so we allow the request to go through, which will yield an empty response
+                    if (resolvedIndices.isNoIndicesPlaceholder()) {
+                        putTransientIfNonExisting(AuthorizationServiceField.INDICES_PERMISSIONS_KEY,
+                            IndicesAccessControl.ALLOW_NO_INDICES);
+                        // check action name
+                        authorizeIndexActionName(authentication, action, requestId, unwrappedRequest, authzInfo, authzEngine, listener);
+                    } else {
+                        authorizeIndexAction(authentication, action, requestId, unwrappedRequest, authzInfo,
+                            authzEngine, resolvedIndices, metaData, authorizedIndices, listener);
+                    }
+                }, listener::onFailure));
             } else {
-                // we do this here in addition to the denial below since we might run into an assertion on scroll request below if we
-                // don't have permission to read cross cluster but wrap a scroll request.
-                throw denial(auditId, authentication, action, request, permission.names());
+                authzEngine.authorizeIndexActionName(authentication, unwrappedRequest, action, authzInfo,
```

Comment:
I don't understand why on this branch we authorize first on the action name.

Reply:
This is to preserve the preexisting behavior; if we don't it changes the return/error in some cases. The existing block of code that triggered the branch is this one:

https://github.com/elastic/elasticsearch/blob/891a37adaa4804dd2cc5fc060d864bb29e593efe/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L286-L290

Ultimately if we do not allow remote indices then the user needs the permission to execute the action on at least one index. Since this is the non remote branch, we check the action name first.

The branch above has the nested call to authorize index action name, which represents this block of code in the current AuthorizationService:

https://github.com/elastic/elasticsearch/blob/891a37adaa4804dd2cc5fc060d864bb29e593efe/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L299-L303

Reply:
Gotcha, thank you for the explanation.

Reply:
I **really** appreciate the thorough review, you taking this as an opportunity to learn, and questioning what you see.

I'd eventually like to see these two branches combined in order to simplify this. Given the intent to backport this to 6.x, I don't want to do that with this change but maybe after this feature lands we can re-evaluate the behavior.